### ISSUE TITLE
Another SimpleCache instance uses the folder ... TrackPlayer

### ISSUE THREAD

**Issue by jasongrishkoff** [doublesymmetry/react-native-track-player#296]
On the android-redesign branch, I'm bumping into this error:

```
10-10 19:33:03.219  1845  1845 E AndroidRuntime: java.lang.IllegalStateException: Another SimpleCache instance uses the folder: /data/user/0/com.xxx.xxx/cache/TrackPlayer
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.google.android.exoplayer2.upstream.cache.SimpleCache.<init>(SimpleCache.java:125)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.google.android.exoplayer2.upstream.cache.SimpleCache.<init>(SimpleCache.java:112)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.google.android.exoplayer2.upstream.cache.SimpleCache.<init>(SimpleCache.java:84)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.guichaguri.trackplayer.service.models.Track.toMediaSource(Track.java:148)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.guichaguri.trackplayer.service.player.ExoPlayback.add(ExoPlayback.java:74)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.guichaguri.trackplayer.module.MusicModule.lambda$add$2$MusicModule(MusicModule.java:191)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.guichaguri.trackplayer.module.MusicModule$$Lambda$2.run(Unknown Source:8)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:789)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:98)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:164)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:6541)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
10-10 19:33:03.219  1845  1845 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

**Comment by jasongrishkoff** [comment]
Thanks for fixing, @dcvz

**Comment by Jayiitb** [comment]
@jasongrishkoff I am finding this same error when using current `dev`(branch). Any idea, on what is causing this error?

**Comment by Guichaguri** [comment]
@Jayiitb does it happen in 2075ceb17b671220fde1f23dafcbaed188e87922?

**Comment by Jayiitb** [comment]
@Guichaguri Thanks. The problem flew away. 🛸💨

**Comment by malikoum94** [comment]
Hi,

We are facing this issue again on the new beta Kotlin release any tips or idea to avoid this?

**Comment by PaulCombal** [comment]
Same issue on 3.0.0 after hot reloading

**Comment by VinceBT** [comment]
Look if you are not calling any function on the TrackPlayer class before it is initialized, for me I was doing `updateNowPlayingMetadata` before it had a chance to be initialized

**Comment by shukla-prateek-tft** [comment]
Can you help me out to resolve this issue.

**Comment by AlirezaHadjar** [comment]
Still facing this issue in 3.1.0

**Comment by billouboq** [comment]
Same still having the issue

**Comment by billouboq** [comment]
Fixed it by using this :

```js
TrackPlayer.isServiceRunning()
.then((isRunning) => {
	// calling setupPlayer if it's already set makes app crash
	if (isRunning) {
		return Promise.resolve();
	}

	return TrackPlayer.setupPlayer();
})
```

It was crashing because reload was calling setupPlayer again while it was already running

**Comment by rborn** [comment]
@billouboq will test and let you know ❤️

**Comment by billouboq** [comment]
It does not solved my problem properly so by tweaking around found what seems to be the best solution for me :

```js
const setupPromise = TrackPlayer.setupPlayer()
.then(() => {
	return TrackPlayer.updateOptions({
		stopWithApp: true,
		forwardJumpInterval: 10,
		backwardJumpInterval: 10,
		jumpInterval: 10,
		capabilities: [
			Capability.Play,
			Capability.Pause,
			Capability.JumpForward,
			Capability.JumpBackward,
			Capability.SeekTo,
		],
		compactCapabilities: [
			Capability.Play,
			Capability.Pause,
			Capability.JumpForward,
			Capability.JumpBackward,
		],
	});
})
.catch((err) => {
	if (
		err?.message?.includes(
			'The player has already been initialized via setupPlayer',
		)
	) {
		return Promise.resolve();
	}

	return Promise.reject(err);
});
```
		
Catching this specific error otherwise throwing an error

**Comment by izyspania** [comment]
Same problem here with 3.1.0, did this for now:

 !TrackPlayer.isServiceRunning && await TrackPlayer.setupPlayer(),

**Comment by tnghia944** [comment]
Still facing this issue in 3.1.0

**Comment by AlirezaHadjar** [comment]
None of the mentioned solutions works for me.

**Comment by izyspania** [comment]
use the react-native-track-player@next , its fixed there.

**Comment by VinceBT** [comment]
I am trying this piece of code currently:

```
if (module.hot) {
  module.hot.accept(() => {
    console.log('Reload detected')
    TrackPlayer.destroy();
  })
}
```

**Comment by billouboq** [comment]
Isn’t destroy method removed in version 3 ?

**Comment by tnghia944** [comment]
> use react-native-track-player@next , its fixed there.

can you share me current version ?

**Comment by izyspania** [comment]
> > use react-native-track-player@next , its fixed there.
> 
> can you share me current version ?
Seems that it works with  "version": "2.2.0-rc3",

**Comment by jspizziri** [comment]
Before calling `setupPlayer` you should check that the player is not already set up. The example app shows how to properly do this.

https://github.com/doublesymmetry/react-native-track-player/blob/main/example/src/services/SetupService.ts#L6

With that said, I do think this solution feels a bit hacky and perhaps we should consider building it into the `setupPlayer` itself.

If everyone here could confirm that their problem is solved by doing something similar to the example project I'll change this to a feature request and review it with the core team.

**Comment by jspizziri** [comment]
Actually, I do think the `SimpleCache` issue is slightly different from the setupPlayer in the example app, as the hot-reload crash is reproducible in the example app as it exists currently. So something isn't working quite right.

**Comment by mpivchev** [comment]
This is happening because Metro is still holding reference to some parts of RNTP. A temporary solution is to restart Metro. I will try and find a fix for this.

**Comment by jspizziri** [comment]
The SimpleCache issue has been fixed here https://github.com/doublesymmetry/KotlinAudio/pull/53

Once this is merged and released, we just need to bump the version of KA in RNTP.

**Comment by billouboq** [comment]
That's wonderful thanks 🙏

**Comment by izyspania** [comment]
Nice, thanks.

**Comment by mmmoussa** [comment]
Experienced this same issue when using https://github.com/avishayil/react-native-restart for handling app localization changes. I am guessing https://github.com/doublesymmetry/KotlinAudio/pull/53 will fix it but will test it out once it is released.

**Comment by mmmoussa** [comment]
Figured out how to test this without a new release by modifying the KA dependency in RNTP to `implementation 'com.github.DoubleSymmetry:KotlinAudio:main-SNAPSHOT'` and confirmed that the issue is indeed fixed. Thanks @mpivchev and @jspizziri!

**Comment by VinceBT** [comment]
@mmmoussa If you want you can use https://www.npmjs.com/package/patch-package to keep the modification after reinstalling while waiting for the release

**Comment by mmmoussa** [comment]
@VinceBT already doing that, thanks for the suggestion!

**Comment by izyspania** [comment]
Tested with com.github.DoubleSymmetry:KotlinAudio:main-SNAPSHOT and as @mmmoussa said it works without any problems.

**Comment by Johan-dutoit** [comment]
👋  What's the ETA on the next release?

**Comment by Johan-dutoit** [comment]
@jspizziri Please see my question above.  Anything the community can help with to get this over the line?

**Comment by jspizziri** [comment]
@Johan-dutoit currently the core team is focused elsewhere. I believe this will change in a few weeks. We'd like the next patch release to contain a few other fixes apart from this one. For now, using `patch-package` is probably the best solution.

**Comment by RamProg** [comment]
Is this fix included in the last nightly version? I'm using the one from a month ago and this issue persists.

**Comment by minhchienwikipedia** [comment]
@jspizziri what's we need to change? can you show the detail?

**Comment by HyopeR** [comment]
Hello. Roadmap for those who want to apply the fix while waiting for the update.

1- `yarn add -D patch-package postinstall-postinstall` or `npm i --save-dev patch-package postinstall-postinstall`
2- Add "prepare": "patch-package" under scripts in your package.json file. It will look like this.
```
"scripts": {
    "prepare": "patch-package",
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
},
```
3- From your editor go to node_modules/react-native-track-player/android/build.gradle
4- Find this part: implementation 'com.github.DoubleSymmetry:KotlinAudio:v1.2.4'
5- Replace with: implementation 'com.github.DoubleSymmetry:KotlinAudio:main-SNAPSHOT'
6- In terminal go to the project root directory.
7- `yarn patch-package react-native-track-player` or `npm patch-package react-native-track-player`

**Comment by Johan-dutoit** [comment]
and now everyone using the patch is in a world of pain...because main snapshot moved on.

**Comment by mmmoussa** [comment]
Patches should be using this:
`implementation 'com.github.DoubleSymmetry:KotlinAudio:dd4bdab147c90fffcff7800790dbfcb41291d29b'`

**Comment by Naturalclar** [comment]
Seems like using `implementation 'com.github.DoubleSymmetry:KotlinAudio:main-SNAPSHOT` is no longer working due to another PR being merged to KotlinAudio which adds breaking change to the package (https://github.com/doublesymmetry/KotlinAudio/pull/50)

**Comment by jspizziri** [comment]
Alright, I'm going to lock this issue since there is no value to be added at this point. I'll reiterate the where everything is at one more time:

1. This issue has been fixed (see above)
2. The fix has **NOT** been released yet (so it's not available in `3.2.0`, please don't expect it to be)
3. The fix **WILL** be released in the next version of RNTP which is being worked on, but progress is slow.
4. The workaround, for now, is stated above, you can use `patch-package` to get around it.

I encourage you to become a code contributor and help us move issues across the board more quickly. That will help you get the outcome you desire more quickly. If you want to know how to become a contributor you can read the README or reach out to me!
